### ISSUE TITLE
Series.unique() has some kind of typing issue

### ISSUE THREAD

**Issue by Dr-Irv** [microsoft/python-type-stubs#65]
Pyright 2021.5.3
```python
import pandas as pd

df = pd.DataFrame([[10, 1, "x"], [20, 0, "z"], [30, 1, "x"]], columns=["a", "b", "xz"])

f = df["xz"]

reveal_type(f)

xzl = f.unique()

reveal_type(xzl)
```

- The type of `f` is reported as `Type of "f" is "Series[Dtype@__getitem__]"`
- The type of `xzl` is reported as `Unknown`

Looking at the type stubs, I can't figure out why `xzl` comes out as `Unknown`, so this might be a pyright issue, not a type stub issue.  I can't find an obvious fix in the type stubs.  The type stubs for `Series` has:
```python
def unique(self) -> np.ndarray: ...
```
so I don't understand why the type of `xzl` is `Unknown`

**Comment by jakebailey** [comment]
Hm, I can't seem to reproduce this:

![image](https://user-images.githubusercontent.com/5341706/119550182-f1fd3700-bd4c-11eb-93eb-b35f2d668952.png)

**Comment by Dr-Irv** [comment]
@jakebailey are you running with the release 2021.5.3 ?  Here's my pic:
![image](https://user-images.githubusercontent.com/15113894/119551612-d3f10000-bd67-11eb-8016-c3f7edb0ec22.png)

**Comment by jakebailey** [comment]
Yep, I am. Is your selected interpreter one without numpy, by chance?

**Comment by Dr-Irv** [comment]
No.  It has `numpy` .  Here is output of command-line `pyright`:
```
No configuration file found.
No pyproject.toml file found.
stubPath c:\Code\Misc\pytypes\typings is not a valid directory.
Assuming Python platform Windows
Searching for source files
Found 1 source file
c:\Code\Misc\pytypes\getitemreset.py
  c:\Code\Misc\pytypes\getitemreset.py:8:13 - info: Type of "f" is "Unknown"
  c:\Code\Misc\pytypes\getitemreset.py:12:13 - info: Type of "xzl" is "Unknown"
0 errors, 0 warnings, 2 infos
Completed in 1.503sec
```

**Comment by Dr-Irv** [comment]
What's odd is that my colleague doesn't have this problem.   Same versions, etc.

**Comment by jakebailey** [comment]
When you go-to-declaration on `unique`, where does it take you?

**Comment by Dr-Irv** [comment]
Figured it out.  I had "Use library implementations to extract type information when type stub is not present" unchecked because I have other 3rd party libraries that are not typed and I want to ignore errors in VSC when those libraries are used.

Which brings me back to this enhancement request I put in 8 months ago for `pylance`: https://github.com/microsoft/pylance-release/issues/379

**Comment by Dr-Irv** [comment]
Although, I should add that I get the same issue when it is checked and I have numpy 1.20.x installed.

**Comment by jakebailey** [comment]
I guess I don't know why that option would be required; the pandas type stubs only refer to numpy, and if you have 1.20+ it should be using those types even without useLibraryCodeForTypes enabled. I'd be interested to see if you could just use `np.array` in your test case and run pyright on that without `--lib`.

**Comment by Dr-Irv** [comment]
I'm not sure how to modify the test case to use `np.array` since it is all pandas code!

My point is this:
1. If I use a conda environment with numpy < 1.20, and have "Use library implementations to extract type information when type stub is not present" checked,  I get the expected behavior (`xzl` is `np.ndarray`)
2. If I use a conda environment with numpy < 1.20, and have "Use library implementations to extract type information when type stub is not present" *not* checked,  I get the wrong behavior (`xzl` is `Unknown`)
3. If I use a conda environment with numpy >= 1.20,  I get the wrong behavior (`xzl` is `Unknown`)  whether or not "Use library implementations to extract type information when type stub is not present" is  checked,

**Comment by jakebailey** [comment]
> I'm not sure how to modify the test case to use np.array since it is all pandas code!

I mean to test doing:

```python
import numpy as np

reveal_type(np.ndarray)
```

Because I can't reproduce your third item, which makes me think that there's something going on with numpy.

**Comment by jakebailey** [comment]
The answer to "When you go-to-declaration on `unique`, where does it take you?" would also be useful.

**Comment by Dr-Irv** [comment]
For the 3 cases, I listed above, here is the result of the `numpy` test you suggested:

1. If I use a conda environment with numpy < 1.20, and have "Use library implementations to extract type information when type stub is not present" checked, 

  -  I get  that `np.array` is `"(object: Unknown, dtype: Unknown = ..., *, copy: Unknown = ..., order: Unknown = ..., subok: Unknown = ..., ndmin: Unknown = ..., like: Unknown = ...) -> Any"`
  -  If I follow the definition of `unique()`, it goes to the pandas file in that environment.
  - If I follow the definition of `np.array`, it goes to the `pyi` file in the pylance distribution

2. If I use a conda environment with numpy < 1.20, and have "Use library implementations to extract type information when type stub is not present" not checked, 

  -  I get that `np.array` is `Unknown`
  -  If I follow the definition of `unique()`, it goes to the pandas file in that environment.
  - I cannot follow the definition of np.array (No definition found)

3. If I use a conda environment with numpy >= 1.20, independent of the setting of "Use library implementations to extract type information when type stub is not present":

  -  I get the result `Type of "np.array" is "(object: object, dtype: DTypeLike = ..., *, copy: bool = ..., order: _OrderKACF = ..., subok: bool = ..., ndmin: int = ..., like: ArrayLike = ...) -> ndarray"`
  - I cannot follow the definition of `unique` because it is on the symbol `f` which is `Unknown`
  - The defnition of `np.array` goes to the numpy installation in that environment.

**Comment by jakebailey** [comment]
I think you did go-to-definition, but the think I'm looking for is go-to-declaration, which would tell me where the stubs are coming from. That's what I'm mainly interested in here, chasing around the stubs to figure out where the types are coming from and where the difference in our setups is.

It should be the cast that with the third setup, you can go-to-declaration in the pandas stub folder in pylance, then to the py.typed numpy library.

**Comment by Dr-Irv** [comment]
Following the declarations...

1. If I use a conda environment with numpy < 1.20, and have "Use library implementations to extract type information when type stub is not present" checked, 
   - `unique` goes to `series.pyi` in the pylance distribution
   - `np.ndarray` from `series.pyi` goes to `__init__.py` in the numpy distribution in the environment
   - `np.array` from test source goes to `_multiarray_umath.pyi` from the pylance distribution

2. If I use a conda environment with numpy < 1.20, and have "Use library implementations to extract type information when type stub is not present" not checked, 
   - `unique` goes to `series.pyi` in the pylance distribution
   - `np.ndarray` from `series.pyi` has no declaration
   - No declaration for `np.array` in test source

3. If I use a conda environment with numpy >= 1.20, independent of the setting of "Use library implementations to extract type information when type stub is not present":
  - No declaration for `unique()`
  - `np.array` goes to `__init__.pyi` in the numpy distribution

**Comment by Dr-Irv** [comment]
It really seems that the issue is the setting of ""Use library implementations to extract type information when type stub is not present"   With an environment where numpy < 1.20, if that setting is off then pyright isn't finding the `numpy` type stubs that I thought pylance supplied.  Here's a simple example:
```python
import numpy as np
from typing import Union


def myfun(a: Union[float, np.ndarray]):
    if isinstance(a, float):
        return 0
    else:
        return len(a)


def myfun2(a: Union[float, np.ndarray]):
    reveal_type(a)
    if isinstance(a, np.ndarray):
        return len(a)
    else:
        return 0
```

In `myfun2`, with that setting off, and numpy 1.19.2, pyright complains about `len(a)` being invalid, because it doesn't know that `np.ndarray` is a type.

**Comment by jakebailey** [comment]
I'm going to transfer this to the Pylance tracker, since I don't think this is actually an issue with the stubs here.

But it's definitely the case that <1.20, pyright isn't going to get the generated stubs that ship with Pylance. They are a different kind of stub that's only used when we are trying to analyze a compiled module and shouldn't be used in type checking (as the info is inaccurate and mainly for the large bulk of users who just want editor features). If you want to type check your code and use numpy, 1.20 is going to be a must.
